OpenRA.Mods.Common/Widgets/Logic/IrcLogic.cs

```diff
@@ -0,0 +1,316 @@
+
+			nicknameBox = widget.Get<TextFieldWidget>("NICKNAME_BOX");
+			nicknameBox.Text = Game.Settings.Irc.Nickname;
+			nicknameBox.OnTextEdited = () =>
+			{
+				nicknameBox.Text = SanitizedName(nicknameBox.Text);
+				Game.Settings.Irc.Nickname = nicknameBox.Text;
+				Game.Settings.Save();
+			};
+
+			connectBG = widget.Get("IRC_CONNECT_BG");
+			ircContainer = widget.Get("IRC_CONTAINER");
+
+			var disconnectButton = widget.Get<ButtonWidget>("DISCONNECT_BUTTON");
+			disconnectButton.IsDisabled = () => client == null;
+			disconnectButton.OnClick = Disconnect;
+
+			MaybeShowConnectPanel();
+		}
+
+		static string SanitizedName(string dirty)
+		{
+			if (string.IsNullOrEmpty(dirty))
+				return null;
+
+			// TODO: some special chars are allowed as well, but not at every position
+			var clean = new string(dirty.Where(c => char.IsLetterOrDigit(c)).ToArray());
+
+			if (string.IsNullOrEmpty(clean))
+				return null;
+
+			if (char.IsDigit(clean[0]))
+				return SanitizedName(clean.Substring(1));
+
+			// Source: https://tools.ietf.org/html/rfc2812#section-1.2.1
+			if (clean.Length > 9)
+				clean = clean.Substring(0, 9);
+
+			return clean;
+		}
+
+		void MaybeShowConnectPanel()
+		{
+			if (client != null && client.IsConnected)
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+				return;
+			}
+
+			if (Game.Settings.Irc.ConnectAutomatically)
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+				Connect();
+				return;
+			}
+
+			ircContainer.Visible = false;
+			connectBG.Visible = true;
+
+			var connectAutomaticallyCheckBox = connectBG.Get<CheckboxWidget>("CONNECT_AUTOMATICALLY_CHECKBOX");
+			connectAutomaticallyCheckBox.IsChecked = () => Game.Settings.Irc.ConnectAutomatically;
+			connectAutomaticallyCheckBox.OnClick = () => Game.Settings.Irc.ConnectAutomatically ^= true;
+
+			var connectButton = connectBG.Get<ButtonWidget>("CONNECT_BUTTON");
+			connectButton.OnClick = () =>
+			{
+				ircContainer.Visible = true;
+				connectBG.Visible = false;
+
+				Game.Settings.Irc.ConnectAutomatically = connectAutomaticallyCheckBox.IsChecked();
+				Game.Settings.Save();
+				Connect();
+			};
+		}
+
+		void Connect()
+		{
+			client = new IrcClient();
+			client.Encoding = System.Text.Encoding.UTF8;
+			client.SendDelay = 100;
+			client.ActiveChannelSyncing = true;
+
+			client.OnConnected += OnConnected;
+			client.OnError += OnError;
+			client.OnRawMessage += OnRawMessage;
+			client.OnJoin += OnJoin;
+			client.OnChannelActiveSynced += OnChannelActiveSynced;
+			client.OnNickChange += OnNickChange;
+			client.OnPong += OnPong;
+
+			try
+			{
+				AddChatLine("Connecting to {0}...".F(Game.Settings.Irc.Hostname));
+				client.Connect(Game.Settings.Irc.Hostname, Game.Settings.Irc.Port);
+			}
+			catch (Exception e)
+			{
+				AddChatLine("Connection error: {0}".F(e.Message));
+				Log.Write("irc", e.ToString());
+			}
+
+			listenThread = new Thread(new ThreadStart(Listen));
```

Comment:
Can you give the thread a name - I find them handy when distinguishing between threads during debugging.